zkevm-circuits/src/evm_circuit/execution/returndatasize.rs

```diff
@@ -0,0 +1,186 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::N_BYTES_MEMORY_ADDRESS,
+        step::ExecutionState,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition::Delta},
+            from_bytes, CachedRegion, RandomLinearCombination,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    table::CallContextFieldTag,
+    util::Expr,
+};
+use bus_mapping::evm::OpcodeId;
+use eth_types::{Field, ToLittleEndian};
+use halo2_proofs::plonk::Error;
+
+#[derive(Clone, Debug)]
+pub(crate) struct ReturnDataSizeGadget<F> {
+    same_context: SameContextGadget<F>,
+    return_data_size: RandomLinearCombination<F, N_BYTES_MEMORY_ADDRESS>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ReturnDataSizeGadget<F> {
+    const NAME: &'static str = "RETURNDATASIZE";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::RETURNDATASIZE;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        // Add lookup constraint in the call context for the returndatasize field.
+        let return_data_size = cb.query_rlc();
+        cb.call_context_lookup(
+            false.expr(),
+            None,
+            CallContextFieldTag::LastCalleeReturnDataLength,
+            from_bytes::expr(&return_data_size.cells),
+        );
+
+        // The returndatasize should be pushed to the top of the stack.
+        cb.stack_push(return_data_size.expr());
+
+        let step_state_transition = StepStateTransition {
+            rw_counter: Delta(2.expr()),
+            program_counter: Delta(1.expr()),
+            stack_pointer: Delta((-1).expr()),
+            gas_left: Delta(-OpcodeId::RETURNDATASIZE.constant_gas_cost().expr()),
+            ..Default::default()
+        };
+
+        let same_context = SameContextGadget::construct(cb, opcode, step_state_transition);
+
+        Self {
+            same_context,
+            return_data_size,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        _call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+        let return_data_size = block.rws[step.rw_indices[1]].stack_value();
+        self.return_data_size.assign(
+            region,
+            offset,
+            Some(
+                return_data_size.to_le_bytes()[..N_BYTES_MEMORY_ADDRESS]
```

Comment:
```suggestion
                return_data_size.to_le_bytes()[..N_BYTES_U64]
```

And the corresponding change here as well

Reply:
Updated.

Reply:
ditto